File: src/services/sanctioningAuthorityService.ts
import { authService } from './authService';

export interface SanctioningAuthorityResponse {
  sanctioningAuthorityCode: string;
  sanctioningAuthorityFullName: string;
  sanctioningAuthorityShortName: string;
  hierarchyOrder: number;
  fromDate: string;
  toDate: string | null;
  userId: string;
  regStatus: string;
  regTime: string | null;
  active: boolean;
}

export interface SanctioningAuthorityRequest {
  sanctioningAuthorityCode: string;
  sanctioningAuthorityFullName: string;
  sanctioningAuthorityShortName: string;
  hierarchyOrder: number;
  fromDate: string;
  toDate: string | null;
  userId: string;
  regStatus: string;
}

const API_BASE_URL = "http://localhost:7080/api";

class SanctioningAuthorityService {
  private getHeaders(): Record<string, string> {
    const authHeaders = authService.getAuthHeader();
    
    return {
      "Content-Type": "application/json",
      ...authHeaders
    };
  }

  async getAllSanctioningAuthorities(): Promise<SanctioningAuthorityResponse[]> {
    try {
      console.log('Fetching sanctioning authorities from:', `${API_BASE_URL}/sanctioning-authorities`);
      const response = await fetch(`${API_BASE_URL}/sanctioning-authorities`, {
        method: "GET",
        headers: this.getHeaders(),
      });

      console.log('Response status:', response.status);

      if (!response.ok) {
        let errorMessage = `HTTP ${response.status}: Failed to fetch sanctioning authorities`;
        try {
          const errorData = await response.json();
          if (errorData.message) {
            errorMessage = errorData.message;
          }
        } catch (e) {
          // Response is not JSON, use default message
        }
        throw new Error(errorMessage);
      }

      const data = await response.json();
      console.log('Sanctioning authorities loaded:', data);
      return data;
    } catch (error) {
      const errorMsg = error instanceof Error ? error.message : 'Unknown error occurred';
      console.error("Error fetching sanctioning authorities:", errorMsg);
      throw new Error(errorMsg || 'Failed to fetch sanctioning authorities');
    }
  }

  async getActiveSanctioningAuthorities(): Promise<SanctioningAuthorityResponse[]> {
    try {
      const response = await fetch(`${API_BASE_URL}/sanctioning-authorities/active`, {
        method: "GET",
        headers: this.getHeaders(),
      });

      if (!response.ok) {
        let errorMessage = `HTTP ${response.status}: Failed to fetch active sanctioning authorities`;
        try {
          const errorData = await response.json();
          if (errorData.message) {
            errorMessage = errorData.message;
          }
        } catch (e) {
          // Response is not JSON
        }
        throw new Error(errorMessage);
      }

      return await response.json();
    } catch (error) {
      const errorMsg = error instanceof Error ? error.message : 'Unknown error occurred';
      console.error("Error fetching active sanctioning authorities:", errorMsg);
      throw new Error(errorMsg || 'Failed to fetch active sanctioning authorities');
    }
  }

  async getSanctioningAuthorityByCode(code: string): Promise<SanctioningAuthorityResponse> {
    try {
      const response = await fetch(`${API_BASE_URL}/sanctioning-authorities/${code}`, {
        method: "GET",
        headers: this.getHeaders(),
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `HTTP ${response.status}: Failed to fetch sanctioning authority`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error fetching sanctioning authority:", error);
      throw error;
    }
  }

  async createSanctioningAuthority(request: SanctioningAuthorityRequest): Promise<SanctioningAuthorityResponse> {
    try {
      const response = await fetch(`${API_BASE_URL}/sanctioning-authorities`, {
        method: "POST",
        headers: this.getHeaders(),
        body: JSON.stringify(request),
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `HTTP ${response.status}: Failed to create sanctioning authority`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error creating sanctioning authority:", error);
      throw error;
    }
  }

  async updateSanctioningAuthority(code: string, request: SanctioningAuthorityRequest): Promise<SanctioningAuthorityResponse> {
    try {
      const response = await fetch(`${API_BASE_URL}/sanctioning-authorities/${code}`, {
        method: "PUT",
        headers: this.getHeaders(),
        body: JSON.stringify(request),
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `HTTP ${response.status}: Failed to update sanctioning authority`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error updating sanctioning authority:", error);
      throw error;
    }
  }

  async deactivateSanctioningAuthority(code: string): Promise<void> {
    try {
      const response = await fetch(`${API_BASE_URL}/sanctioning-authorities/${code}/deactivate`, {
        method: "PUT",
        headers: this.getHeaders(),
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `HTTP ${response.status}: Failed to deactivate sanctioning authority`);
      }
    } catch (error) {
      console.error("Error deactivating sanctioning authority:", error);
      throw error;
    }
  }

  async deleteSanctioningAuthority(code: string): Promise<void> {
    try {
      const response = await fetch(`${API_BASE_URL}/sanctioning-authorities/${code}`, {
        method: "DELETE",
        headers: this.getHeaders(),
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `HTTP ${response.status}: Failed to delete sanctioning authority`);
      }
    } catch (error) {
      console.error("Error deleting sanctioning authority:", error);
      throw error;
    }
  }
}

export const sanctioningAuthorityService = new SanctioningAuthorityService();

File: pms-backend/src/main/java/com/pms/service/SanctioningAuthorityService.java
package com.pms.service;

import com.pms.dto.SanctioningAuthorityRequest;
import com.pms.dto.SanctioningAuthorityResponse;
import com.pms.entity.SanctioningAuthority;
import com.pms.repository.SanctioningAuthorityRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class SanctioningAuthorityService {

    @Autowired
    private SanctioningAuthorityRepository repository;

    // Get all sanctioning authorities
    public List<SanctioningAuthorityResponse> getAllSanctioningAuthorities() {
        return repository.findAllOrdered().stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }

    // Get all active sanctioning authorities
    public List<SanctioningAuthorityResponse> getAllActiveSanctioningAuthorities() {
        return repository.findAllActive().stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }

    // Get all inactive sanctioning authorities
    public List<SanctioningAuthorityResponse> getAllInactiveSanctioningAuthorities() {
        return repository.findAllInactive().stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }

    // Get sanctioning authority by code
    public SanctioningAuthorityResponse getSanctioningAuthorityByCode(String code) {
        return repository.findBySanctioningAuthorityCode(code)
                .map(this::convertToResponse)
                .orElse(null);
    }

    // Create sanctioning authority
    public SanctioningAuthorityResponse createSanctioningAuthority(SanctioningAuthorityRequest request) {
        if (repository.existsBySanctioningAuthorityCode(request.getSanctioningAuthorityCode())) {
            throw new IllegalArgumentException("Sanctioning Authority Code already exists");
        }

        SanctioningAuthority authority = SanctioningAuthority.builder()
                .sanctioningAuthorityCode(request.getSanctioningAuthorityCode())
                .sanctioningAuthorityFullName(request.getSanctioningAuthorityFullName())
                .sanctioningAuthorityShortName(request.getSanctioningAuthorityShortName())
                .hierarchyOrder(request.getHierarchyOrder())
                .fromDate(request.getFromDate())
                .toDate(request.getToDate())
                .userId(request.getUserId())
                .regStatus(request.getRegStatus() != null ? request.getRegStatus() : "A")
                .regTime(LocalDate.now())
                .build();

        SanctioningAuthority saved = repository.save(authority);
        return convertToResponse(saved);
    }

    // Update sanctioning authority
    public SanctioningAuthorityResponse updateSanctioningAuthority(String code, SanctioningAuthorityRequest request) {
        SanctioningAuthority authority = repository.findBySanctioningAuthorityCode(code)
                .orElseThrow(() -> new IllegalArgumentException("Sanctioning Authority not found"));

        authority.setSanctioningAuthorityFullName(request.getSanctioningAuthorityFullName());
        authority.setSanctioningAuthorityShortName(request.getSanctioningAuthorityShortName());
        authority.setHierarchyOrder(request.getHierarchyOrder());
        authority.setFromDate(request.getFromDate());
        authority.setToDate(request.getToDate());
        authority.setUserId(request.getUserId());
        authority.setRegStatus(request.getRegStatus());

        SanctioningAuthority updated = repository.save(authority);
        return convertToResponse(updated);
    }

    // Deactivate sanctioning authority
    public void deactivateSanctioningAuthority(String code) {
        SanctioningAuthority authority = repository.findBySanctioningAuthorityCode(code)
                .orElseThrow(() -> new IllegalArgumentException("Sanctioning Authority not found"));
        
        authority.setRegStatus("I");
        authority.setToDate(LocalDate.now());
        repository.save(authority);
    }

    // Delete sanctioning authority
    public void deleteSanctioningAuthority(String code) {
        if (repository.existsBySanctioningAuthorityCode(code)) {
            repository.deleteById(code);
        } else {
            throw new IllegalArgumentException("Sanctioning Authority not found");
        }
    }

    // Convert entity to response
    private SanctioningAuthorityResponse convertToResponse(SanctioningAuthority entity) {
        return SanctioningAuthorityResponse.builder()
                .sanctioningAuthorityCode(entity.getSanctioningAuthorityCode())
                .sanctioningAuthorityFullName(entity.getSanctioningAuthorityFullName())
                .sanctioningAuthorityShortName(entity.getSanctioningAuthorityShortName())
                .hierarchyOrder(entity.getHierarchyOrder())
                .fromDate(entity.getFromDate())
                .toDate(entity.getToDate())
                .userId(entity.getUserId())
                .regStatus(entity.getRegStatus())
                .regTime(entity.getRegTime())
                .build();
    }
}
